ta/pkcs11/src/processing_symm.c

```diff
@@ -751,6 +857,73 @@ enum pkcs11_rc step_symm_operation(struct pkcs11_session *session,
 
 		break;
 
+	case PKCS11_CKM_MD5_HMAC_GENERAL:
+	case PKCS11_CKM_SHA_1_HMAC_GENERAL:
+	case PKCS11_CKM_SHA224_HMAC_GENERAL:
+	case PKCS11_CKM_SHA256_HMAC_GENERAL:
+	case PKCS11_CKM_SHA384_HMAC_GENERAL:
+	case PKCS11_CKM_SHA512_HMAC_GENERAL:
+		assert(session->processing->extra_ctx);
+		hmac_len = *(uint32_t *)session->processing->extra_ctx;
+
+		switch (function) {
+		case PKCS11_FUNCTION_SIGN:
+			if (out_size < hmac_len) {
+				/* inform client of required size */
+				out_size = hmac_len;
+				output_data = true;
+				rc = PKCS11_CKR_BUFFER_TOO_SMALL;
+				goto out;
+			}
+
+			res = TEE_MACComputeFinal(proc->tee_op_handle,
+						  in_buf, in_size,
+						  computed_mac,
+						  &computed_mac_size);
+			if (res == TEE_SUCCESS) {
+				/* truncate to hmac_len */
+				memcpy(out_buf, computed_mac, hmac_len);
+			}
+			if (res == TEE_SUCCESS || res == TEE_ERROR_SHORT_BUFFER)
+				output_data = true;
+
+			/* inform client of required size */
```

Comment:
As for logic goes -- there was already step before with exactly same comment to verify and report back required size. Eg. check should be done for size.

If we get here we have more or less copied the data there already.

You could get rid of the comment altogether if you move setting of out_size next to memory copying spot.

Now this comment should reflect that we just tell what we copied?

Reply:
input value for `out_size` may be higher than `hmac_len` and here we are expected to return the effective output data size. The implementation looks good to me.